internal/protocol/connector.go

```diff
@@ -116,11 +116,13 @@ func (c *Connector) Connect(ctx context.Context) (*Protocol, error) {
 // Make a single attempt to establish a connection to the leader server trying
 // all addresses available in the store.
 func (c *Connector) connectAttemptAll(ctx context.Context, log logging.Func) (*Protocol, error) {
-	servers, err := c.store.Get(ctx)
+	storedServers, err := c.store.Get(ctx)
 	if err != nil {
 		return nil, errors.Wrap(err, "get servers")
 	}
-
+	// Create a copy of the stored servers to get around the data race
+	servers := make([]NodeInfo, len(storedServers))
+	copy(servers, storedServers)
```

Comment:
How do you stop the race occurring here?

Reply:
I would have thought this would need to be covered by a mutex for all operations of c.store.Get()

Reply:
The data race is actually in the slice sort and it appears to be caused by `Set` being called on the store in the middle of the sort so operating on a copy of the slice fixes that. The lock does not fix the issue because the call to `Set` is done outside of the connector which will allow the slice to still be modified in the middle of the sort.

Reply:
Thanks for the clarification. Does the race still exist though if the `storedServers` is modified during `copy`?

Reply:
I ran the test that was triggering it with `-count 1000` and could not reproduce it. Without the `copy` it reproduced on 90% of runs with `-count 1` and 100% of the time if I had `-count 2`

Reply:
OK, its certainly an improvement, and won't prevent this PR from being merged, but I don't think its fully solved it.

Reply:
Ya, likely a better way to solve it is to have a lock in the store and make a copy in the get method and return that copy